### PR TITLE
tstest/integration: add test for root TUN mode (including race mode)

### LINKED ISSUE #9801: race: magicsock.Conn.populateCLIPingResponseLocked vs localapi.Handler.servePing
Some in-progress integration tests that run tailscaled as root and in race mode just found this race:

```
    integration_test.go:938: WARNING: DATA RACE
    integration_test.go:938: Read at 0x00c0003296c0 by goroutine 258:
    integration_test.go:938:   reflect.Value.Float()
    integration_test.go:938:       reflect/value.go:1391 +0xb3
    integration_test.go:938:   encoding/json.floatEncoder.encode()
    integration_test.go:938:       encoding/json/encode.go:531 +0xc0
    integration_test.go:938:   encoding/json.floatEncoder.encode-fm()
    integration_test.go:938:       <autogenerated>:1 +0x86
    integration_test.go:938:   encoding/json.structEncoder.encode()
    integration_test.go:938:       encoding/json/encode.go:706 +0x2bd
    integration_test.go:938:   encoding/json.structEncoder.encode-fm()
    integration_test.go:938:       <autogenerated>:1 +0xe4
    integration_test.go:938:   encoding/json.ptrEncoder.encode()
    integration_test.go:938:       encoding/json/encode.go:878 +0x373
    integration_test.go:938:   encoding/json.ptrEncoder.encode-fm()
    integration_test.go:938:       <autogenerated>:1 +0x86
    integration_test.go:938:   encoding/json.(*encodeState).reflectValue()
    integration_test.go:938:       encoding/json/encode.go:323 +0x83
    integration_test.go:938:   encoding/json.(*encodeState).marshal()
    integration_test.go:938:       encoding/json/encode.go:295 +0x184
    integration_test.go:938:   encoding/json.(*Encoder).Encode()
    integration_test.go:938:       encoding/json/stream.go:209 +0x137
    integration_test.go:938:   tailscale.com/ipn/localapi.(*Handler).servePing()
    integration_test.go:938:       tailscale.com/ipn/localapi/localapi.go:1414 +0x4ee
    integration_test.go:938:   tailscale.com/ipn/localapi.(*Handler).ServeHTTP()
    integration_test.go:938:       tailscale.com/ipn/localapi/localapi.go:192 +0x80b
    integration_test.go:938:   tailscale.com/ipn/ipnserver.(*Server).serveHTTP()
    integration_test.go:938:       tailscale.com/ipn/ipnserver/server.go:205 +0xfd1
    integration_test.go:938:   tailscale.com/ipn/ipnserver.(*Server).serveHTTP-fm()
    integration_test.go:938:       <autogenerated>:1 +0x51
    integration_test.go:938:   net/http.HandlerFunc.ServeHTTP()
    integration_test.go:938:       net/http/server.go:2136 +0x47
    integration_test.go:938:   net/http.serverHandler.ServeHTTP()
    integration_test.go:938:       net/http/server.go:2938 +0x2a1
    integration_test.go:938:   net/http.(*conn).serve()
    integration_test.go:938:       net/http/server.go:2009 +0xc24
    integration_test.go:938:   net/http.(*Server).Serve.func3()
    integration_test.go:938:       net/http/server.go:3086 +0x4f
    integration_test.go:938: 
    integration_test.go:938: Previous write at 0x00c0003296c0 by goroutine 55:
    integration_test.go:938:   tailscale.com/wgengine/magicsock.(*Conn).populateCLIPingResponseLocked()
    integration_test.go:938:       tailscale.com/wgengine/magicsock/magicsock.go:765 +0xc7
    integration_test.go:938:   tailscale.com/wgengine/magicsock.(*endpoint).handlePongConnLocked()
    integration_test.go:938:       tailscale.com/wgengine/magicsock/endpoint.go:1111 +0x12ce
    integration_test.go:938:   tailscale.com/wgengine/magicsock.(*Conn).handleDiscoMessage.func1()
    integration_test.go:938:       tailscale.com/wgengine/magicsock/magicsock.go:1426 +0x84
    integration_test.go:938:   tailscale.com/wgengine/magicsock.(*peerMap).forEachEndpointWithDiscoKey()
    integration_test.go:938:       tailscale.com/wgengine/magicsock/peermap.go:120 +0x159
    integration_test.go:938:   tailscale.com/wgengine/magicsock.(*Conn).handleDiscoMessage()
    integration_test.go:938:       tailscale.com/wgengine/magicsock/magicsock.go:1425 +0xf29
    integration_test.go:938:   tailscale.com/wgengine/magicsock.(*Conn).receiveIP()
    integration_test.go:938:       tailscale.com/wgengine/magicsock/magicsock.go:1192 +0x1ca
    integration_test.go:938:   tailscale.com/wgengine/magicsock.(*connBind).Open.(*Conn).receiveIPv4.(*Conn).mkReceiveFunc.func1()
    integration_test.go:938:       tailscale.com/wgengine/magicsock/magicsock.go:1165 +0x504
    integration_test.go:938:   github.com/tailscale/wireguard-go/device.(*Device).RoutineReceiveIncoming()
    integration_test.go:938:       github.com/tailscale/wireguard-go@v0.0.0-20230929223258-2f6748dc88e7/device/receive.go:111 +0x528
    integration_test.go:938:   github.com/tailscale/wireguard-go/device.(*Device).BindUpdate.func3()
    integration_test.go:938:       github.com/tailscale/wireguard-go@v0.0.0-20230929223258-2f6748dc88e7/device/device.go:532 +0x4f
    integration_test.go:938: 
    integration_test.go:938: Goroutine 258 (running) created at:
    integration_test.go:938:   net/http.(*Server).Serve()
    integration_test.go:938:       net/http/server.go:3086 +0x80c
    integration_test.go:938:   tailscale.com/ipn/ipnserver.(*Server).Run()
    integration_test.go:938:       tailscale.com/ipn/ipnserver/server.go:521 +0x69a
    integration_test.go:938:   main.startIPNServer()
    integration_test.go:938:       tailscale.com/cmd/tailscaled/tailscaled.go:460 +0x791
    integration_test.go:938:   main.run()
    integration_test.go:938:       tailscale.com/cmd/tailscaled/tailscaled.go:399 +0x6ba
    integration_test.go:938:   main.main()
    integration_test.go:938:       tailscale.com/cmd/tailscaled/tailscaled.go:242 +0xc69
    integration_test.go:938: 
    integration_test.go:938: Goroutine 55 (running) created at:
    integration_test.go:938:   github.com/tailscale/wireguard-go/device.(*Device).BindUpdate()
    integration_test.go:938:       github.com/tailscale/wireguard-go@v0.0.0-20230929223258-2f6748dc88e7/device/device.go:532 +0x61e
    integration_test.go:938:   github.com/tailscale/wireguard-go/device.(*Device).upLocked()
    integration_test.go:938:       github.com/tailscale/wireguard-go@v0.0.0-20230929223258-2f6748dc88e7/device/device.go:172 +0x47
    integration_test.go:938:   github.com/tailscale/wireguard-go/device.(*Device).changeState()
    integration_test.go:938:       github.com/tailscale/wireguard-go@v0.0.0-20230929223258-2f6748dc88e7/device/device.go:153 +0xf9
    integration_test.go:938:   github.com/tailscale/wireguard-go/device.(*Device).Up()
    integration_test.go:938:       github.com/tailscale/wireguard-go@v0.0.0-20230929223258-2f6748dc88e7/device/device.go:210 +0x23b3
    integration_test.go:938:   tailscale.com/wgengine.NewUserspaceEngine()
    integration_test.go:938:       tailscale.com/wgengine/userspace.go:417 +0x2391
    integration_test.go:938:   main.tryEngine()
    integration_test.go:938:       tailscale.com/cmd/tailscaled/tailscaled.go:679 +0x8e4
    integration_test.go:938:   main.createEngine()
    integration_test.go:938:       tailscale.com/cmd/tailscaled/tailscaled.go:577 +0x248
    integration_test.go:938:   main.getLocalBackend()
    integration_test.go:938:       tailscale.com/cmd/tailscaled/tailscaled.go:484 +0x244
    integration_test.go:938:   main.startIPNServer.func2()
    integration_test.go:938:       tailscale.com/cmd/tailscaled/tailscaled.go:450 +0x344
```

/cc @maisem @raggi

### PR DESCRIPTION
**Pull request by bradfitz**
Two commits:

* fix a race in magicsock (found by second commit)
* add integration test that runs as root in TUN mode & catches races (with @maisem)

Fixes #9801
Updates #7894

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -90,6 +90,8 @@ jobs:
         sudo apt-get -y install qemu-user
     - name: build test wrapper
       run: ./tool/go build -o /tmp/testwrapper ./cmd/testwrapper
+    - name: integration test (early)
+      run: PATH=$PWD/tool:$PATH /tmp/testwrapper ./tstest/integration/ ${{matrix.buildflags}}
```
Comment (bradfitz) [inline .github/workflows/test.yml:94]: this is temporary just to make CI get to the good stuff first

File: tstest/integration/integration_test.go [hunk 2]
```diff
@@ -1143,10 +1179,18 @@ func (n *testNode) AwaitListening() {
...
 		c, err := safesocket.Connect(s)
-		if err != nil {
-			return err
+		if err == nil {
+			c.Close()
+			return nil
+		}
+		// If we're running a tun-mode test requiring sudo, we might
+		// not have access from this non-root context to hit the server.
+		// So also try a slower way:
```
Comment (maisem) [inline tstest/integration/integration_test.go:1188]: is this still right?

Comment (maisem) [reply tstest/integration/integration_test.go:1188]: the test needs to run as root in order to run tailscaled as root?

File: wgengine/magicsock/endpoint.go [hunk 3]
```diff
@@ -475,9 +474,20 @@ func (de *endpoint) noteActiveLocked() {
...
+	taken atomic.Bool // first CompareAndSwamp from false to true takes ownership of res
+	res   *ipnstate.PingResult
+	cb    func(*ipnstate.PingResult)
+}
+
+func (p *pingResultAndCallback) reply() bool {
+	return p != nil && p.taken.CompareAndSwap(false, true)
+}
+
+// discoPing starts a disc-level ping for the "tailscale ping" command (or other
```
Comment (maisem) [inline wgengine/magicsock/endpoint.go:487]: Disco-level

File: wgengine/magicsock/endpoint.go [hunk 4]
```diff
@@ -475,9 +474,20 @@ func (de *endpoint) noteActiveLocked() {
...
+// callers, such as c2n). res is value to call cb with, already partially
+// filled. cb must be called at most once. Once called, ownership of res passes to db.
```
Comment (maisem) [inline wgengine/magicsock/endpoint.go:489]: Passes to de?

Comment (maisem) [reply wgengine/magicsock/endpoint.go:489]: Or passes to cb?

**Review by maisem — APPROVED** [review]

**Pull request merged**